plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/PartitionPreprocessor.java

```diff
@@ -9,107 +9,84 @@
-            .collect(Collectors.toList());
-        
-        // For each pair of consecutive partition IDs (a, b), connect all nodes from partition a to partition b
-        Iterator<Integer> idIterator = sortedPartitionIDs.iterator();
-        
-        Integer firstId = idIterator.next();
-        while (idIterator.hasNext()) {
-            Integer secondId = idIterator.next();
-            connectNodes(partitionToNodesMap.get(firstId), partitionToNodesMap.get(secondId));
-            firstId = secondId;
-        }
+            .filter(lNode -> lNode.hasProperty(LayeredOptions.PARTITIONING_PARTITION))
+            .flatMap(lNode -> Streams.stream(lNode.getOutgoingEdges()))
+            .filter(lEdge -> mustBeReversed(lEdge))
+            .collect(Collectors.toList())
+            
+            .stream()
+            .forEach(lEdge -> reverse(lEdge, lGraph));
         
         monitor.done();
     }
 
-    /**
-     * Connects all nodes from the first collection to all nodes from the second collection.
-     */
-    private void connectNodes(final Collection<LNode> firstPartition, final Collection<LNode> secondPartition) {
-        for (LNode node : firstPartition) {
-            LPort sourcePort = new LPort();
-            sourcePort.setNode(node);
-            sourcePort.setSide(PortSide.EAST);
-            sourcePort.setProperty(InternalProperties.PARTITION_DUMMY, true);
+    private boolean mustBeReversed(final LEdge lEdge) {
+        assert lEdge.getSource().getNode().hasProperty(LayeredOptions.PARTITIONING_PARTITION);
+        
+        if (lEdge.getTarget().getNode().hasProperty(LayeredOptions.PARTITIONING_PARTITION)) {
+            Integer sourcePartition = lEdge.getSource().getNode().getProperty(LayeredOptions.PARTITIONING_PARTITION);
+            Integer targetPartition = lEdge.getTarget().getNode().getProperty(LayeredOptions.PARTITIONING_PARTITION);
             
-            for (LNode otherNode : secondPartition) {
-                LPort targetPort = new LPort();
-                targetPort.setNode(otherNode);
-                targetPort.setSide(PortSide.WEST);
-                targetPort.setProperty(InternalProperties.PARTITION_DUMMY, true);
-                
-                LEdge edge = new LEdge();
-                edge.setProperty(InternalProperties.PARTITION_DUMMY, true);
-                edge.setProperty(LayeredOptions.PRIORITY_DIRECTION, PARTITION_CONSTRAINT_EDGE_PRIORITY);
-                edge.setSource(sourcePort);
-                edge.setTarget(targetPort);
-            }
+            // We need to reverse an edge if sourcePartition > targetPartition
+            return sourcePartition.compareTo(targetPartition) > 0;
+            
+        } else {
+            return false;
         }
     }
 
+    private void reverse(final LEdge lEdge, final LGraph lGraph) {
+        lEdge.reverse(lGraph, true);
+        lEdge.setProperty(LayeredOptions.PRIORITY_DIRECTION, PARTITION_CONSTRAINT_EDGE_PRIORITY);
```

Comment:
We could check if a non-default priority is already set, and, if so, add it to the constant. Might help a user to manually "fix" an issue when nodes without partitions are present.